File: streamlit_app/lib/trends.py
"""Pricing trends over time, by market — ported from trends.js. Avg sale
$/acre and avg lease $/SF (net + gross), bucketed monthly, alongside the
BoC overnight/approx-prime rate for the same months.
"""

import datetime as dt

from . import db
from .rates import get_monthly_rates


def _months_back(n, end_month):
    y, m = int(end_month[:4]), int(end_month[5:7])
    months = []
    for i in range(n - 1, -1, -1):
        mm = m - i
        yy = y
        while mm <= 0:
            mm += 12
            yy -= 1
        months.append(f"{yy:04d}-{mm:02d}")
    return months


def _avg(arr):
    return sum(arr) / len(arr) if arr else None


def get_trends(months_count=12):
    rows = [dict(r) for r in db.query("SELECT * FROM listings WHERE date_flagged = 0 AND record_date IS NOT NULL")]
    latest = max((r["record_date"] for r in rows), default="0000-00")
    end_month = latest[:7] if latest != "0000-00" else dt.datetime.now().strftime("%Y-%m")
    months = _months_back(months_count, end_month)

    markets = sorted({r["market"] for r in rows})
    by_market = {}
    for mkt in markets:
        market_rows = [r for r in rows if r["market"] == mkt]
        sale_per_acre, lease_net_psf, lease_gross_psf, txn_count = [], [], [], []
        for month in months:
            in_month = [r for r in market_rows if r["record_date"][:7] == month]
            sales = [r for r in in_month if r["price_deal_type"] == "sale" and r["price_unit"] == "total" and r["acres"] and r["price_amount"]]
            lease_net = [r for r in in_month if r["price_deal_type"] == "lease" and r["price_unit"] == "per_sf" and r["price_basis"] == "net"]
            lease_gross = [r for r in in_month if r["price_deal_type"] == "lease" and r["price_unit"] == "per_sf" and r["price_basis"] == "gross"]
            sale_per_acre.append(_avg([r["price_amount"] / r["acres"] for r in sales]))
            lease_net_psf.append(_avg([r["price_amount"] for r in lease_net]))
            lease_gross_psf.append(_avg([r["price_amount"] for r in lease_gross]))
            txn_count.append(sum(1 for r in in_month if r["record_type"] == "transaction"))
        by_market[mkt] = {"sale_per_acre": sale_per_acre, "lease_net_psf": lease_net_psf, "lease_gross_psf": lease_gross_psf, "txn_count": txn_count}

    rate_by_month = {r["month"]: r for r in get_monthly_rates()}
    rates = [rate_by_month.get(m, {"month": m, "overnight_rate": None, "approx_prime": None}) for m in months]

    return {"months": months, "by_market": by_market, "rates": rates}

File: server/src/trends.js
'use strict';

const { db } = require('./db');
const { getMonthlyRates } = require('./rates');

/**
 * Pricing trends over time, by market — avg sale $/acre and avg lease $/SF
 * (net + gross), bucketed monthly, plus the BoC overnight/approx-prime rate
 * for the same months so the two can be overlaid on one chart.
 */

function monthsBack(n, endMonth) {
  const months = [];
  const [y, m] = endMonth.split('-').map(Number);
  const d = new Date(Date.UTC(y, m - 1, 1));
  for (let i = n - 1; i >= 0; i--) {
    const dd = new Date(d);
    dd.setUTCMonth(dd.getUTCMonth() - i);
    months.push(dd.toISOString().slice(0, 7));
  }
  return months;
}

function avg(arr) {
  return arr.length ? arr.reduce((a, b) => a + b, 0) / arr.length : null;
}

function getTrends(monthsCount = 12) {
  // date_flagged rows excluded — see pulse.js/pricing.js for why (the "2206" typo).
  const rows = db.prepare("SELECT * FROM listings WHERE date_flagged = 0 AND record_date IS NOT NULL").all();
  const latest = rows.reduce((max, r) => (r.record_date > max ? r.record_date : max), '0000-00');
  const endMonth = latest.slice(0, 7) || new Date().toISOString().slice(0, 7);
  const months = monthsBack(monthsCount, endMonth);

  const markets = [...new Set(rows.map((r) => r.market))];
  const byMarket = {};
  for (const mkt of markets) {
    const marketRows = rows.filter((r) => r.market === mkt);
    const salePerAcre = [];
    const leaseNetPsf = [];
    const leaseGrossPsf = [];
    const txnCount = [];
    for (const month of months) {
      const inMonth = marketRows.filter((r) => r.record_date.slice(0, 7) === month);
      const sales = inMonth.filter((r) => r.price_deal_type === 'sale' && r.price_unit === 'total' && r.acres && r.price_amount);
      const leaseNet = inMonth.filter((r) => r.price_deal_type === 'lease' && r.price_unit === 'per_sf' && r.price_basis === 'net');
      const leaseGross = inMonth.filter((r) => r.price_deal_type === 'lease' && r.price_unit === 'per_sf' && r.price_basis === 'gross');
      salePerAcre.push(avg(sales.map((r) => r.price_amount / r.acres)));
      leaseNetPsf.push(avg(leaseNet.map((r) => r.price_amount)));
      leaseGrossPsf.push(avg(leaseGross.map((r) => r.price_amount)));
      txnCount.push(inMonth.filter((r) => r.record_type === 'transaction').length);
    }
    byMarket[mkt] = { salePerAcre, leaseNetPsf, leaseGrossPsf, txnCount };
  }

  const rateByMonth = new Map(getMonthlyRates().map((r) => [r.month, r]));
  const rates = months.map((m) => rateByMonth.get(m) || { month: m, overnightRate: null, approxPrime: null });

  return { months, byMarket, rates };
}

module.exports = { getTrends };
